상품 API 구현

jgm0327 opened the pull request:
- productEntity 추가
- 상품 API를 위한 예외 처리
- 메인 페이지를 위해 cache 설정
- 응답과 요청을 위한 Dto
- 상품 API 경로를 위해 SecurityConfig, JwtFilter 수정
- 설정 파일에 잘못된 부분 수정

jgm0327 committed:
Fix: MemberEntity, SecurityConfig 수정

- MemberEntity의 enum String으로 저장하도록 수정

- MemberEntity의 id를 memberId로 수정

- memberId로 수정하고 SecurityConfig에서 getId()를 getMemberId()로 수정

Resolves: #10

jgm0327 committed:
Fix: MemberService 수정

- findById(1L)를 findByEmail()로 수정하고 예외도 MemberNotFoundException로 수정

Resolves: #10

jgm0327 committed:
Fix: MemberController 수정

- http header에 토큰 담을 때 Bearer 를 추가

Resolves: #10

jgm0327 committed:
Feat: productEntity 생성 및 memberEntity와 매핑

- 판매여부를 위한 ProductStatus를 만들고 ProductEntity를 생성

- MemberEntity에 매핑 객체 만들고 addProduct로 매핑되로고 구현

- ProductRepository 생성

Resolves: #9

jgm0327 committed:
Feat: 상품 API 예외 처리

- 상품 API에서 발생할 수 있는 예외들 클래스 생성

- ProductErrorCode로 에러 설정

Resolves: #9

jgm0327 committed:
Feat: Cache 설정

- Caffeine Cache와 AbstractCache를 위한 dependency 추가

- Cache Config와 Cache Type으로 Cache 설정

- Main함수에 @EnableCaching을 붙여서 캐시 이용 가능하도록 수정

Resolves: #9

jgm0327 committed:
Feat: Dto 생성

- 각 요청 및 응답에 대한 Dto 생성

- ProductMapper를 통해 Dto, Entity 사이에 변환

- 페이지네이션 정보를 담아서 응답 객체를 보내줄 PageResultDto 생성

- seller는 따로 String 형식으로 추가하기 위해 ignore로 설정

Resolves: #9

jgm0327 committed:
Feat: ProductService 생성

- 전체 조회에서 Page별로 Cache의 키 값으로 하기 위해 KeyGenerator 생성

- MemberService에서 findById() 메소드를 만들고 ProductService에서 이용할 수 있도록 함.

Resolves: #9

jgm0327 committed:
Feat: interceptor 추가

- 상품 조회 시 존재하는 상품인지 interceptor에서 확인

- 상품 수정 및 삭제를 할 때 판매자와 같은 사용자인지 확인을 interceptor에서 확인

- WebMvcConfig에 interceptor 추가

Resolves: #9

jgm0327 committed:
Feat: ProductController 추가

Resolves: #9

jgm0327 committed:
Feat: 보안 수정

- 상품 API를 위한 경로에 대한 필터링을 위해 jwtFilter와 SecurityConfig에 추가

- 토큰 시간 길게 수정

Resolves: #9

jgm0327 committed:
Fix: 설정 파일 수정

Resolves: #9

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/common/cache/CacheType.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+public enum CacheType {
+    Main_PAGE("mainPage", 1, 1000);
```
```suggestion
    MAIN_PAGE("mainPage", 1, 1000);
```

보통 열거형 상수는 대문자를 쓰는것이 컨벤션이더라구요.
https://stackoverflow.com/questions/3069743/coding-conventions-naming-enums

jgm0327 replied:
PAGE는 대문자로 하고 Main에서는 이렇게 하다니 손가락에 힘이 풀렸나봐요

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/common/cache/CacheType.java:
```diff
@@ -0,0 +1,18 @@
+public enum CacheType {
+    Main_PAGE("mainPage", 1, 1000);
+
+    CacheType(String cacheName, int expireAfterWrite, int maximumSize) {
+        this.cacheName = cacheName;
+        this.expireAfterWrite = expireAfterWrite;
+        this.maximumSize = maximumSize;
+    }
```
해당 cache type은 expireAfterWrite, maximumSize 만 설정할 수 있는거 같아요.
아래와 같이 accessAfterWrite 등도 다른 방향도 설정할수 있게끔 하는 방법도 있을거 같구, ExpireAfterWriteCacheType이라고 명시적으로 cache type을 한정해주는것도 좋을거 같아요.

```java
@Getter
public enum CacheType {
    MAIN_PAGE("mainPage", 1, 1000, ExpireType.EXPIRE_AFTER_WRITE),
    DETAIL_PAGE("detailPage", 2, 500, ExpireType.ACCESS_AFTER_WRITE);

    private final String cacheName;
    private final int expireDuration;
    private final int maximumSize;
    private final ExpireType expireType;

    CacheType(String cacheName, int expireDuration, int maximumSize, ExpireType expireType) {
        this.cacheName = cacheName;
        this.expireDuration = expireDuration;
        this.maximumSize = maximumSize;
        this.expireType = expireType;
    }

    public boolean isExpireAfterWrite() {
        return expireType == ExpireType.EXPIRE_AFTER_WRITE;
    }

    public boolean isAccessAfterWrite() {
        return expireType == ExpireType.ACCESS_AFTER_WRITE;
    }

    public enum ExpireType {
        EXPIRE_AFTER_WRITE,
        ACCESS_AFTER_WRITE
    }
}
```

jgm0327 replied:
이런식으로도 작성할 수 있겠네요. 좋은 인사이트 얻어 가요!! 👍

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/DanggunMarketApplication.java:
```diff
@@ -2,10 +2,12 @@
 
 @SpringBootApplication
 @EnableJpaAuditing
+@EnableCaching
```
요게 두군데 들어간거 같네요ㅎㅎ

jgm0327 replied:
저번에 key 설정 안 될 때, 저거 작성하라고 해서 했다가 안 지웠나보네요

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/common/config/WebMvcConfig.java:
```diff
@@ -0,0 +1,23 @@
+public class WebMvcConfig implements WebMvcConfigurer {
+    private final ValidProductIdInterceptor validProductIdInterceptor;
+    private final ValidAuthorizeProductInterceptor validAuthorizeProductInterceptor;
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(validProductIdInterceptor)
+                .addPathPatterns("/v1/products/**");
+
+        registry.addInterceptor(validAuthorizeProductInterceptor)
+                .addPathPatterns("/v1/products/**");
+    }
```
이거 전 회사에서 여러개일때 관리했던 방식이 생각났는데 아래와 같이 했던걸로 기억해요.
각 interceptor 별로 url만 별도로 관리할수 있도록.

```java

    private static final String[] aaaInterceptorPath = {
        "/v1/products/**",
        "/v1/item/**"
    };

    private static final String[] bbbInterceptorPath = {
        "/v1/orders/**"
    };

    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(aaaInterceptor)
                .addPathPatterns(aaaInterceptorPath);

        registry.addInterceptor(bbbInterceptor)
                .addPathPatterns(bbbInterceptorPath);
    }
```

jgm0327 replied:
저렇게 하면 인터셉터가 많아져도 관리하기 좀 더 수월하겠네요!

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/product/ProductController.java:
```diff
@@ -0,0 +1,64 @@
+        return ResponseEntity.ok().body(response);
+    }
+
+    @PostMapping("/v1/products/{id}")
```
상품수정 api인데 post 메소드로 잘못 설정하신듯 해요.

jgm0327 replied:
수장하도록 하겠습니다!!

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/product/ProductEntity.java:
```diff
@@ -0,0 +1,58 @@
+
+    private String name;
+    private String description;
+    private int price;
```
규민님 돈의 적절한 자료형에 대해서 멘토링시간에 한번 얘기나눴던걸로 기억해요. 한번 다시 생각부탁드릴게요~!

jgm0327 replied:
중고 거래 사이트 특성상 할인같은 개념은 없다고 생각해서 int 형으로 했는데, 그래도 BigDecimal을 사용하는 게 좋을까요?

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/product/ProductService.java:
```diff
@@ -0,0 +1,113 @@
+    private final ProductRepository productRepository;
+    private final MemberService memberService;
+
+    @Transactional
```
addProductWithLocalImage 메소드에서 어떤 사유로 transactional 어노테이션이 필요하다고 생각하셨는지 궁금합니다.

jgm0327 replied:
member.addProduct를 통해서 ProductEntity와 매핑을 해야 하는데, @Transactional을 붙이지 않으면 DB 세션을 바로 닫아버려서 no Session 에러가 발생하여 필요하다고 생각했습니다.

MemberEntity 안에 ProductEntity가 LAZY로 가져오기 때문에 생기는 문제인 것 같습니다.

f-lab-kane commented on danggunMarket/src/main/java/com/example/danggunmarket/product/ProductService.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Transactional
+    public GetProductDetailResponse editProduct(long id, EditProductRequest request){
```
상품 수정할때 동시성 이슈는 고려할 필요가 없을까요?

jgm0327 replied:
1. 어느 유저의 장바구니 안에 상품이 담겨 있을 때
2. 상품 조회할 때
3. 어느 유저가 상품을 구매하고 있을 때

이러한 상황들을 고려해야 할 것 같네요.

f-lab-kane reviewed (approved):
규민님 수고하셨습니다 👍

jgm0327 reviewed (commented):
코드 리뷰 감사합니다!!

jgm0327 committed:
Fix: ProductErrorCode 수정

- PRODUCT가 아닌 MEMBER로 작성한 거 수정

fix: #9

jgm0327 committed:
Fix: Cache 수정

- CacheType과 CacheConfig을 수정해서 write 뿐만 아니라 access 삭제 정책도 가능하도록 함.

fix: #9

jgm0327 committed:
Fix: @EnableCaching 삭제

- DanggunMarketApplication에서 @EnableCaching 삭제

fix: #9

jgm0327 committed:
Fix: 상품 인터셉터 수정

- PUT과 DELETE에서만 변경 권한 있는지 확인

- UriTemplate를 이용해서 pathVariable 추출하도록 수정

fix: #9

jgm0327 committed:
Fix: GetProductResponse 수정

- mapstruct가 매핑할 수 있도록 id를 productId로 수정

fix: #9

jgm0327 committed:
Fix: ProductService 수정

- ProductErrorCode에 MEMBER라고 선언된 값 변경

fix: #9

jgm0327 committed:
Fix: ProductController 수정

- 수정할 때 PUT 메소드로 요청 받도록 설정

fix: #9
